replica/distributed_loader.cc

```diff
@@ -406,6 +407,22 @@ sstables::shared_sstable make_sstable(replica::table& table, fs::path dir, sstab
     return sstm.make_sstable(table.schema(), dir.native(), generation_value, sstm.get_highest_supported_format(), sstables::sstable_format_types::big, gc_clock::now(), &error_handler_gen_for_upload_dir);
 }
 
+class sstable_generation_generator {
+    // We still want to do our best to keep the generation numbers shard-friendly.
+    // Each destination shard will manage its own generation counter.
+    //
+    // operator() is called by multiple shards in parallel when performing reshard,
+    // so we have to use atomic<> here.
+    std::atomic<int64_t> _last_generation;
```

Comment:
> s/int64_t/sstables::generation_type::int_t/

i understand the rationales behind this. the only thing i don't like is that it's too lengthy: 7 chars versus 32 chars. i will change this commit anyway.

> 
> Or better keep (non-atomic) `sstables::generation_type` and use `submit_to` (or `invoke_on` if we instantiate a `sharded<generation_generator>`) to manipulate it, calling `replica::table::make_new_generation` to generate new generation so we have a single source for this logic.

the idea of `sharded<generation_generator>` was considered in a previous revision of this PR. but it was considered overkill by both Avi and Botond. see https://github.com/scylladb/scylladb/pull/13073#discussion_r1130840951 (i cannot find Avi's comment on this, it was on another PR). and i agree with them that the generation generator should not a sharded service. not because it cannot be, just because the complexity of it does not deserve a new service. i plan to drop  `replica::table::make_new_generation` in a follow up PR, as it is only used by tests where a generation generator would suffice. another reason to drop this static function is, by then, we would need to access `_sstables_manager.feature_service().uuid_sstable_identifiers` when generating a new generation. one way to get around this to pass the generation generator, last known generation and `uuid_sstable_identifiers` to this static method, but this renders this method a very thin wrapper of the generator, and its whole point is to call it with other parameters.

Reply:
> > s/int64_t/sstables::generation_type::int_t/
> 
> i understand the rationales behind this. the only thing i don't like is that it's too lengthy: 7 chars versus 32 chars. i will change this commit anyway.

Hopefully it will go away, or remain in a few select places after the transition to uuid-based sstable generations.

> > Or better keep (non-atomic) `sstables::generation_type` and use `submit_to` (or `invoke_on` if we instantiate a `sharded<generation_generator>`) to manipulate it, calling `replica::table::make_new_generation` to generate new generation so we have a single source for this logic.
> 
> the idea of `sharded<generation_generator>` was considered in a previous revision of this PR. but it was considered overkill by both Avi and Botond. see [#13073 (comment)](https://github.com/scylladb/scylladb/pull/13073#discussion_r1130840951) (i cannot find Avi's comment on this, it was on another PR). and i agree with them that the generation generator should not a sharded service. not because it cannot be, just because the complexity of it does not deserve a new service.

I thought of just using the sharded instance in the distributed loader to get away from std::atomic across shards.
This is not a performance-intensive path so I'd prefer clarity of the code over performance / misleading simplicity.

> i plan to drop `replica::table::make_new_generation` in a follow up PR, as it is only used by tests where a generation generator would suffice. another reason to drop this static function is, by then, we would need to access `_sstables_manager.feature_service().uuid_sstable_identifiers` when generating a new generation. one way to get around this to pass the generation generator, last known generation and `uuid_sstable_identifiers` to this static method, but this renders this method a very thin wrapper of the generator, and its whole point is to call it with other parameters.

okay, looking forward to it :)

Reply:
> > > s/int64_t/sstables::generation_type::int_t/
> > 
> > 
> > i understand the rationales behind this. the only thing i don't like is that it's too lengthy: 7 chars versus 32 chars. i will change this commit anyway.
> 
> Hopefully it will go away, or remain in a few select places after the transition to uuid-based sstable generations.
> 

yeah, we still have some tests where we hardwire the type to integer. but yeah, i think we will fix them if they don't really need to use integer. i think the reason they use integer is just that it was convenient.

> > > Or better keep (non-atomic) `sstables::generation_type` and use `submit_to` (or `invoke_on` if we instantiate a `sharded<generation_generator>`) to manipulate it, calling `replica::table::make_new_generation` to generate new generation so we have a single source for this logic.
> > 
> > 
> > the idea of `sharded<generation_generator>` was considered in a previous revision of this PR. but it was considered overkill by both Avi and Botond. see [#13073 (comment)](https://github.com/scylladb/scylladb/pull/13073#discussion_r1130840951) (i cannot find Avi's comment on this, it was on another PR). and i agree with them that the generation generator should not a sharded service. not because it cannot be, just because the complexity of it does not deserve a new service.
> 
> I thought of just using the sharded instance in the distributed loader to get away from std::atomic across shards. This is not a performance-intensive path so I'd prefer clarity of the code over performance / misleading simplicity.
> 

i like the idea. at first glance, the `atomic<>` is always a signal of bad smell from a scylla developer's perspective. i think we can address this in a follow up change once we all agree on this. @xemul , hi Pavel, what do you think?

> > i plan to drop `replica::table::make_new_generation` in a follow up PR, as it is only used by tests where a generation generator would suffice. another reason to drop this static function is, by then, we would need to access `_sstables_manager.feature_service().uuid_sstable_identifiers` when generating a new generation. one way to get around this to pass the generation generator, last known generation and `uuid_sstable_identifiers` to this static method, but this renders this method a very thin wrapper of the generator, and its whole point is to call it with other parameters.
> 
> okay, looking forward to it :)